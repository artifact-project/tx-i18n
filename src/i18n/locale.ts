let defaultLocale: Locale = {};

export type Locale = {
	[phrase:string]: string;
}

export function setDefaultLocale(local: Locale) {
	defaultLocale = local;
}

export function getTranslate(phrase: string) {
	if (defaultLocale.hasOwnProperty(phrase)) {
		return defaultLocale[phrase];
	}

	// todo: not translated phrase
	return phrase;
}