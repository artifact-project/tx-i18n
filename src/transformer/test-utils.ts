import * as ts from 'typescript';
import { existsSync, readFileSync } from 'fs';
import i18nTxt from './transformer';

const compilerOptions: ts.CompilerOptions = {
    module: ts.ModuleKind.ESNext,
    target: ts.ScriptTarget.ESNext,
    jsx: ts.JsxEmit.Preserve,
    removeComments: false,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
}

export function transform(name: string, options: ts.CompilerOptions = {}) {
    const path = `${__dirname}/fixture/${name}`;
    const fileName = existsSync(`${path}.ts`) ? `${path}.ts` : `${path}.tsx`;
    const output = ts.transpileModule(readFileSync(fileName) + '', {
        fileName: fileName,
        compilerOptions: {
            ...compilerOptions,
            ...options
        },
        transformers: {
            before: [
                i18nTxt({
                    fnName: '__',
                    exclude: [],
                }),
            ],
        },
    });

    return `${output.outputText.trim()}\n`;
}
