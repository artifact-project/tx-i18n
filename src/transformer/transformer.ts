import * as ts from 'typescript';

type HumanTextChecker = (text: string, node: ts.Node) => boolean;

interface Config {
	isHumanText: HumanTextChecker;
	isTranslatableJsxAttribute: (attr: ts.JsxAttribute, element: ts.JsxOpeningLikeElement) => boolean;
	overrideHumanTextChecker?: (isHumanText: HumanTextChecker) => HumanTextChecker;
	fnName: string;
	packageName: string
	fileName: string;
	include: string[];
	exclude: string[];
	compilerOptions: ts.CompilerOptions;
	imports: {
		[name:string]: ts.ImportDeclaration;
	};
}

function log(obj) {
	if (obj == null || /number|string|boolean/.test(typeof obj)) {
		console.log(obj);
		return;
	}

	const copy = {...obj};
	delete copy.parent;
	console.log(copy);
}

const phrases: string[] = [];

function savePhrase(value: string) {
	if (phrases.indexOf(value) === -1) {
		phrases.push(value);
	}

	return value;
}

export function resetPhrases() {
	phrases.length = 0;
}

export function getPhrases() {
	return phrases.sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()));
}

function createImport(name, path) {
	return ts.createImportDeclaration(
		undefined,
		undefined,
		ts.createImportClause(ts.createIdentifier(name), undefined),
		ts.createLiteral(path)
	);
}

function addImport(imports: Config['imports'], name, path) {
	// imports[name] = createImport(name, path);
	// TypeError: Cannot set property text of #<IdentifierObject> which has only a getter

	if (!imports.hasOwnProperty(name)) {
		Object.defineProperty(imports, name, {
			enumerable: true,
			value: createImport(name, path),
		});
	}

	return imports[name];
}

function useNativeImport(module: ts.ModuleKind) {
	return module === ts.ModuleKind.ESNext || module === ts.ModuleKind.ES2015;
}

function i18nId(cfg: Config, prop?: string) {
	const importDecl = addImport(cfg.imports, cfg.fnName, cfg.packageName);
	let id: ts.Expression = useNativeImport(cfg.compilerOptions.module)
		? ts.createIdentifier(cfg.fnName)
		: ts.getGeneratedNameForNode(importDecl);

	if (!useNativeImport(cfg.compilerOptions.module)) {
		id = ts.createPropertyAccess(id, 'default');
	}

	return prop ? ts.createPropertyAccess(id, prop) : id;
}

function visited<T extends ts.Node>(node: T): T {
	node['__visited__'] = true;
	return node;
}

function createLiteral(val: string) {
	const node = ts.createIdentifier(JSON.stringify(val));
	return visited(node);
}

function hasJsxTextChildren(node: ts.JsxElement, cfg: Config) {
	return node.children.some(child => ts.isJsxText(child) && cfg.isHumanText(child.getText(), child));
}

function wrapStringLiteral(node: ts.StringLiteralLike, cfg: Config) {
	const text = node.getText();

	if (!cfg.isHumanText(text, node)) {
		return visited(node);
	}

	savePhrase(text.slice(1, -1));

	return ts.createCall(i18nId(cfg), [], [
		ts.createIdentifier(text),
	])
}

function wrapTemplateExpression(node: ts.TemplateExpression, context, cfg: Config) {
	const args = [];
	let phrase = node.head.text;

	node.templateSpans.forEach((span, idx) => {
		phrase += `<#${(idx + 1)}>${span.literal.text}`;
		args.push(visitNodeAndChildren(span.expression, context, cfg));
	});

	if (!cfg.isHumanText(phrase, node)) {
		return visited(node);
	}

	savePhrase(phrase);

	return ts.createCall(i18nId(cfg), [], [
		createLiteral(phrase),
		ts.createArrayLiteral(args),
	]);
}

function visitNode(node: ts.Node, context, cfg: Config): ts.Node {
	const {
		isHumanText,
	} = cfg;

	if (node['__visited__']) {
		return node;
	}

	if (
		(!node.parent || !ts.isImportDeclaration(node.parent))
		&& (
			ts.isStringLiteral(node)
			|| ts.isNoSubstitutionTemplateLiteral(node)
		)
		&& isHumanText(node.getFullText(), node)
	) {
		const {parent} = node;

		if (parent) {
			if (
				ts.isPropertyAssignment(parent) && parent.name === node
				|| ts.isElementAccessExpression(parent)
			) {
				return node;
			}


			const attr = ts.isJsxAttribute(parent)
				? parent
				: parent.parent && ts.isJsxAttribute(parent.parent) && parent.parent
			;

			if (attr) {
				if (
					!isTagName(attr.parent.parent.tagName.getText())
					|| cfg.isTranslatableJsxAttribute(attr, attr.parent.parent)
				) {
					const newNode = wrapStringLiteral(node, cfg);
					return newNode === node ? node : ts.createJsxExpression(undefined, newNode);
				}
				return node;
			}
		}

		return wrapStringLiteral(node, cfg);
	} else if (ts.isTemplateExpression(node)) {
		const {parent} = node;

		if (parent && (ts.isComputedPropertyName(parent) || ts.isElementAccessExpression(parent))) {
			return node;
		}

		if (
			node.parent
			&& node.parent.parent
			&& ts.isJsxAttribute(node.parent.parent)
			&& (
				!cfg.isTranslatableJsxAttribute(node.parent.parent, node.parent.parent.parent.parent)
				&& isTagName(node.parent.parent.parent.parent.tagName.getText())
			)
		) {
			return node;
		}

		return wrapTemplateExpression(node, context, cfg);
	} else if (ts.isJsxElement(node) && hasJsxTextChildren(node, cfg)) {
		const args = [];
		let simple = true;
		let gpart = 0;
		let phrase = '';
		let hasText = false;

		node.children.forEach(function processing(child) {
			if (ts.isJsxElement(child)) {
				let part = ++gpart;

				simple = false;
				phrase += `<${part}>`;
				args.push(jsxTagToObject(child, context, cfg));
				child.children.forEach(processing)
				phrase += `</${part}>`;
			} else if (ts.isJsxText(child)) {
				hasText = true;
				phrase += child.getFullText().replace(/[\n\t]/g, '');
			} else if (ts.isJsxExpression(child)) {
				simple = false;
				phrase += `<#${++gpart}>`;
				args.push(visitNode(child.expression, context, cfg));
			} else if (ts.isJsxSelfClosingElement(child)) {
				simple = false;
				phrase += `<${++gpart}/>`;
				args.push(jsxTagToObject(child, context, cfg));
			} else {
				log(child);
				throw new Error('Not supproted');
			}
		}, args);

		phrase = phrase.trim();

		if (!hasText || !isHumanText(phrase, node)) {
			return node;
		}

		savePhrase(phrase);

		if (simple) {
			return ts.updateJsxElement(
				node,
				node.openingElement,
				[ts.createJsxExpression(undefined, ts.createCall(
					i18nId(cfg),
					[],
					[].concat(
						createLiteral(phrase),
						args.length ? ts.createArrayLiteral(args) : [],
					),
				))],
				node.closingElement,
			);
		}

		const callExp = ts.createCall(
			i18nId(cfg, 'jsx'),
			[],
			[
				createLiteral(phrase),
				ts.createArrayLiteral([
					jsxTagToObject(node, context, cfg),
					...args,
				]),
			],
		);

		if (ts.isJsxElement(node.parent)) {
			return ts.createJsxExpression(undefined, callExp);
		}

		return callExp;
	}

	return node;
}

function visitNodeAndChildren(node: ts.Node, context, cfg: Config) {
	return ts.visitEachChild(
		visitNode(node, context, cfg),
		(childNode) => visitNodeAndChildren(childNode, context, cfg),
		context,
	);
}

const R_IS_TAG = /^[a-z0-9:]+$/;

function isTagName(name: string) {
	return R_IS_TAG.test(name);
}

function jsxTagToObject(node: ts.JsxElement | ts.JsxSelfClosingElement, context, cfg: Config) {
	const element = ts.isJsxElement(node) ? node.openingElement : node;
	const {attributes} = element;

	let type = element.tagName.getFullText();
	let props: ts.ObjectLiteralElementLike[] = [];

	attributes.properties.forEach(prop => {
		if (ts.isJsxAttribute(prop)) {
			const {initializer} = prop;
			let name = prop.name.getText();
			let value: ts.Expression = null;

			if (initializer) {
				if (cfg.isTranslatableJsxAttribute(prop, element) && cfg.isHumanText(prop.getText(), node)) {
					if (ts.isJsxExpression(initializer)) {
						if (ts.isTemplateExpression(initializer.expression)) {
							value = wrapTemplateExpression(initializer.expression, context, cfg);
						} else {
							value = ts.createIdentifier(initializer.getText().slice(1, -1));
						}
					} else {
						value = wrapStringLiteral(initializer, cfg);
					}
				} else {
					value = ts.isJsxExpression(initializer) ? initializer.expression : initializer;
				}
			}

			props.push(ts.createPropertyAssignment(
				stringifyObjectKey(name),
				value ? value : ts.createNull(),
			));
		} else if (ts.isJsxSpreadAttribute(prop)) {
			// log();
			props.push(ts.createSpreadAssignment(prop.expression));
		}
	});


	return ts.createObjectLiteral([
		ts.createPropertyAssignment(
			'type',
			isTagName(type) ? ts.createIdentifier(`"${type}"`) : element.tagName,
		),
		ts.createPropertyAssignment(
			'props',
			props.length ? ts.createObjectLiteral(props) : ts.createNull(),
		),
	]);
}

function checkPatterFile(this: Config, mask: string) {
	return this.fileName.match(mask) !== null;
}

const R_IS_SIMPLE_OBJ_KEY = /^[a-z\d_]+$/i;
function stringifyObjectKey(name: string) {
	return R_IS_SIMPLE_OBJ_KEY.test(name) ? name : JSON.stringify(name)
}

export type TXConfig = Partial<Pick<Config,
	'fnName'
	| 'packageName'
	| 'include'
	| 'exclude'
	| 'isHumanText'
	| 'isTranslatableJsxAttribute'
	| 'overrideHumanTextChecker'
>>

export default function transformerFactory(config: TXConfig) {
	return function transformer(context: ts.TransformationContext) {
		return function visitor(file: ts.SourceFile) {
			const cfg: Config = {
				fnName: '__',
				packageName: 'tx-i18n',
				fileName: file.fileName,
				exclude: ['/tx-i18n/', '/node_modules/'],
				include: null,
				imports: {},
				compilerOptions: context.getCompilerOptions(),
				isHumanText: (value: string) => /[\wа-яё]/.test(value.trim().replace(/\d+([^\s]+)?/g, '').trim()),
				isTranslatableJsxAttribute: (attr: ts.JsxAttribute) => /^(title|alt|placeholder|value)$/.test(attr.name.getText()),
				...config,
			};

			if (cfg.overrideHumanTextChecker) {
				cfg.isHumanText = cfg.overrideHumanTextChecker(cfg.isHumanText);
			}

			if (cfg.exclude.some(checkPatterFile, cfg)) {
				return file;
			}

			if (cfg.include && !cfg.include.some(checkPatterFile, cfg)) {
				return file;
			}

			try {
				const result = visitNodeAndChildren(file, context, cfg);

				try {
					return ts.updateSourceFileNode(
						result,
						Object.keys(cfg.imports)
							.map(name => cfg.imports[name])
							.concat(result.statements),
					);
				} catch (err) {
					console.error(`\x1b[31m\n[tx-i18n] [update] ${cfg.fileName}\n---\n${err.toString()}\n\x1b[0m`);
				return file;
				}
			} catch (err) {
				console.error(`\x1b[31m\n[tx-i18n] [visit] ${cfg.fileName}\n---\n${err.toString()}\n\x1b[0m`);
				return file;
			}
		}
	}
}
