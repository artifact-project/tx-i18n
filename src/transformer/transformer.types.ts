import * as ts from 'typescript';

export type HumanTextChecker = (text: string, node: ts.Node) => boolean;

export interface Config {
	verbose: (...args: any[]) => void;
	isHumanText: HumanTextChecker;
	normalizeText: (value: string) => string;
	isTranslatableJsxAttribute: (attr: ts.JsxAttribute, element: ts.JsxOpeningLikeElement) => boolean;
	overrideHumanTextChecker?: (isHumanText: HumanTextChecker) => HumanTextChecker;
	textFnName: string;
	jsxFnName: string;
	packageName: string
	fileName: string;
	include: string[];
	exclude: string[];
	pharsesStore?: ContextedPhrases;
	compilerOptions: ts.CompilerOptions;
	imports: {
		[name:string]: ts.ImportDeclaration;
	};
	sourceFile: ts.SourceFile;
};

export type Phrase = {
	value: string;
	file: string;
	ctx: string;
	loc: {
		start: {
			line: number;
			character: number;
		};
		end: {
			line: number;
			character: number;
		};
	};
};

export type ContextedPhrases = {
	[context: string]: Phrase[];
};

export type InlineConfig = {
	context: string;
	normalize: string;
	isJSXExpr: boolean;
};

export type TXConfig = Partial<Pick<Config,
	| 'textFnName'
	| 'jsxFnName'
	| 'packageName'
	| 'include'
	| 'exclude'
	| 'isHumanText'
	| 'isTranslatableJsxAttribute'
	| 'overrideHumanTextChecker'
	| 'normalizeText'
	| 'pharsesStore'
	| 'verbose'
>>;
