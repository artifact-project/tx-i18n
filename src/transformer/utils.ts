import * as ts from 'typescript';
import { Element, Simple } from 'format-message-parse';

export function normJsxElement(node: ts.JsxElement | ts.JsxSelfClosingElement,) {
	return ts.isJsxElement(node) ? node.openingElement : node;
}

export function getJsxTagName(node: ts.JsxElement | ts.JsxSelfClosingElement,) {
	return normJsxElement(node).tagName.getText();
}

export function log(obj: object | string | number | boolean, ind = '', max = 3) {
	if (obj == null || /number|string|boolean/.test(typeof obj)) {
		console.log(obj);
		return;
	}

	const copy = {...Object(obj)};
	let lines = [];
	delete copy['parent'];

	for (let key in copy) {
		lines.push(
			`${ind}${key}: ` + (
			copy[key] === null ? 'null' :
				typeof copy[key] === 'object'
					? max > 0 ? '\n' + log(copy[key], ind + '  ', max - 1) : '{ ... }'
					: copy[key]
		));
	}

	if (ind === '') {
		console.log(lines.join('\n'));
	}

	return lines.join('\n');
}