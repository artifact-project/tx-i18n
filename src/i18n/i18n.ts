import { createElement } from 'react';

import { createCompiler } from './compiler';
import { jsxFactory } from '../jsx/jsx';
import { getTranslate } from './locale';

const compile = createCompiler({
	inject: null,
	partOffset: -1,
	before: () => '',
	after: () => '""',
	text: (text) => `${text}+`,
	value: (part) => `${part}+`,
});

function i18n(phrase: string, parts: any[]) {
	const translatePhrase = getTranslate(phrase);
	const compiledPhrase = compile(translatePhrase);

	return compiledPhrase(parts);
}

i18n['jsx'] = jsxFactory(createElement);

export default i18n as {
	(pharse: string, parts: any[]): string;
	jsx: {
		(pharse: string, parts: any[]): any;
		useWrapper: (Cmp: Function) => void;
	};
};