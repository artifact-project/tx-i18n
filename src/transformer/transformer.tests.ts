import { transform } from './test-utils';
import { getPhrases, resetPhrases } from './transformer';

beforeEach(() => {
	resetPhrases();
})

it('empty', () => {
	expect(transform('empty')).toMatchSnapshot(`'';`);
});

it('numbers', () => {
	expect(transform('numbers')).toMatchSnapshot();
});

it('simple', () => {
	expect(transform('simple')).toMatchSnapshot();
});

it('object', () => {
	expect(transform('object')).toMatchSnapshot();
});

it('template', () => {
	expect(transform('template')).toMatchSnapshot();
});

it('func', () => {
	expect(transform('func')).toMatchSnapshot();
	// expect(getPhrases().default[0].value).toEqual('Single "quote"')
	// expect(getPhrases().default[1].value).toEqual('Double \"quote\"');
	// expect(getPhrases().default[2].value).toEqual('Mixed \"quote\"');
});

it('phrases', () => {
	transform('simple')

	expect(Object.keys(getPhrases())).toEqual(['default']);
	expect(getPhrases().default[0]).toEqual({
		ctx: 'default',
		file: `${__dirname}/fixture/simple.ts`,
		value: 'foo',
		loc: {
			start: {line: 0, character: 0},
			end: {line: 0, character: 5},
		},
	});
});

it('phrases with context', () => {
	resetPhrases();
	transform('context')

	expect(Object.keys(getPhrases())).toEqual([
		'default',
		'other',
	]);

	expect(getPhrases().default.map(p => p.value)).toEqual([
		'foo',
		'bar',
		'qux',
	]);

	expect(getPhrases().other[0].ctx).toEqual('other');
	expect(getPhrases().other.map(p => p.value)).toEqual([
		'bar',
		'foo',
		'bar',
		'baz',
		'qux',
	]);
});
