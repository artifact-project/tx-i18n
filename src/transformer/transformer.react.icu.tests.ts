import { transform, reactLoad, reactRender } from './test-utils';
import { setDefaultLocale, setLocale, setLang } from '../i18n/locale';
import { getPhrases, resetPhrases } from './transformer';
import { plural } from '../../icu/plural/en';
import { plural as ruPlural } from '../../icu/plural/ru';


beforeEach(() => {
	setLang('default')
	resetPhrases();
	setDefaultLocale({}, plural as any);
})

describe('react-icu', () => {
	describe('transform', () => {
		it('Info', () => {
			expect(transform('react-icu')).toMatchSnapshot();
			expect(getPhrases().default.map(v => v.value)).toEqual([
				'You have <a1>{v2, plural, one {# unread message} other {# unread messages}}</a1>.',
				'Total:',
				'{v1, plural, one {# message} other {# messages}}',
			]);
		});
	});

	describe('render', () => {
		const blocks = reactLoad('react-icu');

		it('Info', () => {
			expect(reactRender(blocks.Info, {unread: 1, total: 13})).toMatchSnapshot();
		});

		it('Info: ru', () => {
			setLang('ru');
			setLocale('ru', {
				default: {
					'You have <a1>{v2, plural, one {# unread message} other {# unread messages}}</a1>.': `У вас <a1>{v2, plural,
						one {# непрочитанное сообщение}
						few {# непрочитанных сообщения}
						many {# непрочитанных сообщений}
						other {# непрочитанных сообщений}
					}</a1>.`,
					'Total:': 'Всего:',
					'{v1, plural, one {# message} other {# messages}}': `{v1, plural,
						one {# сообщение}
						few {# сообщения}
						many {# сообщений}
						other {# сообщений}
					}`,
				},
			}, ruPlural);
			expect(reactRender(blocks.Info, {unread: 1, total: 13})).toMatchSnapshot();
		});

		it('Info: ru (partial)', () => {
			setLang('ru');
			setLocale('ru', {
				default: {
					'{v1, plural, one {# message} other {# messages}}': `{v1, plural,
						one {# сообщение}
						few {# сообщения}
						many {# сообщений}
						other {# сообщений}
					}`,
				},
			}, ruPlural);
			expect(reactRender(blocks.Info, {unread: 1, total: 13})).toMatchSnapshot();
		});
	});
});