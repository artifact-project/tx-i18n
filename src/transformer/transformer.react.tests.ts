import { transform, reactLoad, reactRender } from './test-utils';
import { setDefaultLocale } from '../i18n/locale';
import { getPhrases } from './transformer';
import { enPlural } from '../../icu/plural/en';

beforeEach(() => {
	setDefaultLocale({}, enPlural as any);
})

describe('transform', () => {
	it('react', () => {
		expect(transform('react')).toMatchSnapshot();
	});

	it('react-composite', () => {
		expect(transform('react-composite')).toMatchSnapshot();
	});

	it('react-fragments', () => {
		expect(transform('react-fragments')).toMatchSnapshot();
	});
});

describe('render', () => {
	describe('react', () => {
		const blocks = reactLoad('react');

		it('react: Static', () => {
			expect(reactRender(blocks.Static)).toMatchSnapshot();
		});

		it('react: Entities', () => {
			expect(reactRender(blocks.Entities)).toMatchSnapshot();
		});

		it('react: Hello', () => {
			expect(reactRender(blocks.Hello, {username: 'Рубаха'})).toMatchSnapshot();
		});

		it('react: ClickHere', () => {
			expect(reactRender(blocks.ClickHere, {username: 'Рубаха'})).toMatchSnapshot();
		});

		it('react: Fragment', () => {
			expect(reactRender(blocks.Fragment)).toMatchSnapshot();
		});
	});

	describe('react-composite', () => {
		const blocks = reactLoad('react-composite');

		it('react-composite: DeepHello', () => {
			setDefaultLocale({
				default: {
					'Привет {v1}-кун!1': 'Hi, {v1}-kun!',
					'<Hello1/> или Нет!': '<Hello1/> or No!',
				},
			}, enPlural);
			expect(reactRender(blocks.DeepHello, {username: 'i18n'})).toMatchSnapshot();
		});

		it('react-composite: DeepHello (invert)', () => {
			setDefaultLocale({
				default: {
					'Привет {v1}-кун!1': '{v1}, HI!',
					'<Hello1/> или Нет!': 'No or <Hello1/>!',
				},
			}, enPlural);
			expect(reactRender(blocks.DeepHello, {username: 'i18n'})).toMatchSnapshot();
		});

		it('react-composite: Dialog', () => {
			setDefaultLocale({
				default: {
					'Подвал': 'Footer',
					'Хорошо': 'OK',
					'Отмена': 'Cancel',
					'Ширина:': 'Width:',
					'Да{v1}<input2/>': '<input2/> — fail',
					'Тест \"фу\" бар': 'Test "foo" bar',
					'{v1} Окей': 'fail:okey',
				},
				form: {
					'Сохранить': 'Save',
					'<button1/>или<Button2/>': '<button1/> or <Button2/>',
					'Отмена': 'Form cancel',
					'Да{v1}<input2/>': '<input2/> — YES',
					'{v1} Окей': 'Okey {v1}',
				},
			}, enPlural);

			expect(Object.keys(getPhrases())).toEqual(['default', 'form']);
			// expect(getPhrases()).toEqual({});
			expect(reactRender(blocks.Dialog)).toMatchSnapshot();
		});
	});
});