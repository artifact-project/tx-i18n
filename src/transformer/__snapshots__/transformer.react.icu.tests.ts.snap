// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-icu render Info 1`] = `
Array [
  <div
    data-id="unread"
>
    You have 
    <a
        href="#"
    >
        1 unread message
    </a>
    .
</div>,
  <div
    data-id="total"
>
    <label>
        Total:
    </label>
    <b>
        13 messages
    </b>
</div>,
]
`;

exports[`react-icu render Info: ru (partial) 1`] = `
Array [
  <div
    data-id="unread"
>
    You have 
    <a
        href="#"
    >
        1 unread message
    </a>
    .
</div>,
  <div
    data-id="total"
>
    <label>
        Total:
    </label>
    <b>
        13 сообщений
    </b>
</div>,
]
`;

exports[`react-icu render Info: ru 1`] = `
Array [
  <div
    data-id="unread"
>
    У вас 
    <a
        href="#"
    >
        1 непрочитанное сообщение
    </a>
    .
</div>,
  <div
    data-id="total"
>
    <label>
        Всего:
    </label>
    <b>
        13 сообщений
    </b>
</div>,
]
`;

exports[`react-icu transform Info 1`] = `
"import __jsx from \\"tx-i18n/jsx\\";
import __ from \\"tx-i18n/text\\";
import * as React from 'react';
import { enPlural as plural } from '../../../icu/plural/en';
export const Info = ({ unread, total }) => <>
	{__jsx(\\"You have <a1>{v2, plural, one {# unread message} other {# unread messages}}</a1>.\\", [{ type: \\"div\\", props: { \\"data-id\\": \\"unread\\" } }, { type: \\"a\\", props: { href: \\"#\\" } }, { $: plural, _: unread }])}
	<div data-id=\\"total\\">
		<label>{__(\\"Total:\\")}</label>
		<b>{__(\\"{v1, plural, one {# message} other {# messages}}\\", [{ $: plural, _: total }])}</b>
	</div>
</>;
"
`;
